processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java

```diff
@@ -73,302 +60,58 @@ public int size()
   {
-    final int div = Integer.numberOfTrailingZeros(sizePer);
-    final int rem = sizePer - 1;
-    final boolean powerOf2 = sizePer == (1 << div);
-    if(powerOf2) {
-      return new CompressedIndexedLongs() {
-        @Override
-        public long get(int index)
-        {
-          // optimize division and remainder for powers of 2
-          final int bufferNum = index >> div;
-
-          if (bufferNum != currIndex) {
-            loadBuffer(bufferNum);
-          }
-
-          final int bufferIndex = index & rem;
-          return buffer.get(buffer.position() + bufferIndex);
-        }
-      };
-    } else {
-      return new CompressedIndexedLongs();
-    }
+    return supplier.get();
   }
 
   public long getSerializedSize()
   {
-    return baseLongBuffers.getSerializedSize() + 1 + 4 + 4 + 1;
+    return buffer.remaining();
   }
 
   public void writeToChannel(WritableByteChannel channel) throws IOException
   {
-    channel.write(ByteBuffer.wrap(new byte[]{version}));
-    channel.write(ByteBuffer.wrap(Ints.toByteArray(totalSize)));
-    channel.write(ByteBuffer.wrap(Ints.toByteArray(sizePer)));
-    channel.write(ByteBuffer.wrap(new byte[]{compression.getId()}));
-    baseLongBuffers.writeToChannel(channel);
+    channel.write(buffer.asReadOnlyBuffer());
   }
 
   public CompressedLongsIndexedSupplier convertByteOrder(ByteOrder order)
   {
-    return new CompressedLongsIndexedSupplier(
-        totalSize,
-        sizePer,
-        GenericIndexed.fromIterable(baseLongBuffers, CompressedLongBufferObjectStrategy.getBufferForOrder(order, compression, sizePer)),
-        compression
-    );
-  }
-
-  /**
-   * For testing.  Do not use unless you like things breaking
-   */
-  GenericIndexed<ResourceHolder<LongBuffer>> getBaseLongBuffers()
-  {
-    return baseLongBuffers;
+    return fromByteBuffer(buffer, order);
   }
 
   public static CompressedLongsIndexedSupplier fromByteBuffer(ByteBuffer buffer, ByteOrder order)
   {
-    byte versionFromBuffer = buffer.get();
-
-    if (versionFromBuffer == version) {
-      final int totalSize = buffer.getInt();
-      final int sizePer = buffer.getInt();
-      final CompressedObjectStrategy.CompressionStrategy compression = CompressedObjectStrategy.CompressionStrategy.forId(buffer.get());
-      return new CompressedLongsIndexedSupplier(
+    ByteBuffer bufferToUse = buffer.asReadOnlyBuffer();
+    byte versionFromBuffer = bufferToUse.get();
+
+    if (versionFromBuffer == LZF_VERSION || versionFromBuffer == version) {
+      final int totalSize = bufferToUse.getInt();
+      final int sizePer = bufferToUse.getInt();
+      CompressedObjectStrategy.CompressionStrategy compression = CompressedObjectStrategy.CompressionStrategy.LZF;
+      CompressionFactory.LongEncodingFormat encoding = CompressionFactory.DEFAULT_LONG_ENCODING;
+      if (versionFromBuffer == version) {
+        byte compressionId = bufferToUse.get();
+        if (compressionId >= (byte) 0xFE) {
```

Comment:
can we make this >= 0xFE check a method, maybe something like compressionHeaderHasEncoding()